### PR TITLE
Serialize access to the underlying gRPC stream via a single actor

### LINKED ISSUE #14236: Prevent concurrent StreamObserver calls on the same instance
**Describe the bug**

When calling any of the `StreamObserver` methods (including `#onNext`) concurrently, the underlying stream will send garbled messages, and you end up with errors like 

```
java.lang.IllegalStateException: sendHeaders has already been called 
```

Or 

```
io.grpc.StatusRuntimeException: INTERNAL: Connection closed after GOAWAY. HTTP/2 error code: INTERNAL_ERROR, debug data: Stream 7 sent too many headers EOS: false
```

Reading the documentation, it's quite clearly stated in `StreamObserver.java`:

> Separate StreamObservers do not need to be synchronized together; incoming and outgoing directions are independent. Since individual StreamObservers are not thread-safe, if multiple threads will be writing to a StreamObserver concurrently, the application must synchronize calls.

In the `ClientStreamAdapter`, we call `onNext` using a thread pool executor, so concurrent calls are possible.

**To Reproduce**

Easily reproducible with [Camunda 8 Benchmark](https://github.com/camunda-community-hub/camunda-8-benchmark), since there we have a single client, so it's more likely for us to receive all the jobs.

**Expected behavior**

Calls to a `StreamObserver` are serialized. Either all consumers become an actor (what is the overhead there?), or we use gRPC's `SerializingExecutor` to wrap around our thread pool.

**Environment:**
- Zeebe Version: 8.3.0-alpha5, 8.3.0-alpha6

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR refactors the `ClientStreamConsumer` implementation for job streams to serialize access to their `StreamObserver` using actors instead of the internal, unmaintained gRPC class `SerializingExecutor`. The `AsyncJobStreamRemover` had to be made thread safe since it can be accessed by both the actor and the gRPC thread pool, in different order.

This means the stream API is migrated to use actor futures instead of `CompletableFuture`. The old `ClientStreamAdapter` is renamed to `StreamJobHandler` to better reflect what it does, i.e. handle `StreamActivatedJobsRequest` messages. Since it is now an actor, it's created in the `Gateway`, and passed to the `EndpointManager`, much like the long polling job activation handler. This has the nice benefit of more clearly highlighting the inter-class dependencies (i.e. `EndpointManager` does not depend on `Executor` or `ActorSchedulingService`, but on `StreamJobHandler`).

New tests were added, specifically there was no test to handle the case where a stream is _not_ added to the underlying `ClientStreamer`. We now return a `UNAVAILABLE` error, as the only case where a stream is not added right now is if the actor is closed, which would imply a system shutdown (thus retry-able). Previously, the request would hang until the server shuts down, received an error specifying the call was unexpectedly closed.

Finally, we opted to use a single actor to serialize/sequence access instead of multiple actors due to limitations with our actor scheduler, which does not guarantee fairness. Since the `BrokerClient` is a single actor, having a lot of small actors which are busy (i.e. one-actor-per-job-stream) could reduce the throughput of the `BrokerClient`, leading to overall reduced system throughput in some cases.

To avoid too much refactoring here, we'll go with a single actor, which is still generally quite performant. It is expected that we will eventually support better scheduling, and can consider using multiple actors in the future.

## Related issues

closes #14236 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
It seems the single actor performs better than the one-actor-per-stream model. But I'm not very confident in the results, honestly. Max throughput with a single actor is 260 PI/s, and with multiple actors it's 230 PI/s. p99 is worst with single actor, but p50 is similar, and before max throughput was reached, the p99 and p50 of the single actor were better.

You can see from the worker metrics that the single actor workers are activating and handling more jobs as well.

The reason I say I'm not confident is, from the metrics, we can see the multiple actor benchmark, the engines themselves are not pushing as many jobs out...by this I mean, the engine is pushing out ~230 jobs per second, the gateway as well, and the workers are handling them. Whereas the single actor, the engines are pushing out 260 jobs per second...? Commit latency is also better.

Could there be something else affecting the benchmarks?

At any rate, I did a small JMH test to also check this, and it seems like the expected results - more actors can process more messages in less time. But I have a hard time writing such a test without introducing some form of synchronization/contention at the end to await...

Here it is anyway:

<details><summary>MessageThroughputTest.java</summary>

```java
/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
 * one or more contributor license agreements. See the NOTICE file distributed
 * with this work for additional information regarding copyright ownership.
 * Licensed under the Zeebe Community License 1.1. You may not use this file
 * except in compliance with the Zeebe Community License 1.1.
 */
package io.camunda.zeebe.scheduler.perf;

import io.camunda.zeebe.scheduler.Actor;
import io.camunda.zeebe.scheduler.ActorScheduler;
import io.camunda.zeebe.scheduler.future.ActorFutureCollector;
import io.camunda.zeebe.scheduler.future.CompletableActorFuture;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import org.agrona.CloseHelper;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.TearDown;
import org.openjdk.jmh.annotations.Warmup;

@Warmup(iterations = 50, time = 1, batchSize = 1)
@Measurement(iterations = 25, time = 1, batchSize = 1)
@Fork(
    value = 1,
    jvmArgs = {"-Xmx4g", "-Xms4g"})
@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.SECONDS)
@State(Scope.Benchmark)
public class MessageThroughputTest {

  @Benchmark
  public int measureMessageThroughput(final TestState state) throws InterruptedException {
    final var barrier = new CountDownLatch(state.messageCount);
    for (int i = 0; i < state.messageCount; i++) {
      final var actor =
          state.actors.get(ThreadLocalRandom.current().nextInt(0, state.actors.size()));
      actor.send(i, barrier);
    }

    barrier.await();
    return state.messageCount;
  }

  @State(Scope.Benchmark)
  public static class TestState {
    @Param({"1", "10", "100", "1000"})
    private int actorCount;

    @Param({"1000", "10000", "100000", "1000000"})
    private int messageCount;

    private final ActorScheduler scheduler =
        ActorScheduler.newActorScheduler()
            .setCpuBoundActorThreadCount(Runtime.getRuntime().availableProcessors())
            .setIoBoundActorThreadCount(Runtime.getRuntime().availableProcessors() * 2)
            .build();

    private final List<TestActor> actors = new ArrayList<>();

    @Setup
    public void setup() {
      scheduler.start();

      for (int i = 0; i < actorCount; i++) {
        final var actor = new TestActor();
        scheduler.submitActor(actor).join();
        actors.add(actor);
      }
    }

    @TearDown
    public void tearDown() {
      final var actor = Actor.newActor().build();
      final CompletableActorFuture<List<Void>> closed = new CompletableActorFuture<>();
      scheduler.submitActor(actor).join();

      actor.run(
          () -> {
            final var result =
                actors.stream()
                    .map(TestActor::closeAsync)
                    .collect(new ActorFutureCollector<>(actor));
            actor.runOnCompletion(result, closed);
          });
      closed.join();
      CloseHelper.quietCloseAll(actor, scheduler);
      System.gc();
    }
  }

  private static final class TestActor extends Actor {

    private static final Path BLACKHOLE = Path.of("/dev/null");
    private final ByteBuffer buffer = ByteBuffer.allocate(Integer.BYTES);

    private void send(final int base, final CountDownLatch barrier) {
      actor.submit(
          () -> {
            try {
              Files.write(BLACKHOLE, buffer.putInt(0, base).array());
            } catch (final IOException e) {
              throw new RuntimeException(e);
            }

            barrier.countDown();
          });
    }
  }
}
```

</details>

From the results, it seems I'm bounded by threads (since my workload is CPU bounded), so adding more actors than threads doesn't improve things past a certain point :shrug: 

Results with 16 threads (i.e. my # of cores):

```
Benchmark                                       (actorCount)  (messageCount)  Mode  Cnt  Score   Error  Units
MessageThroughputTest.measureMessageThroughput             1            1000    ss   25  0.006 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput             1           10000    ss   25  0.056 ± 0.014   s/op
MessageThroughputTest.measureMessageThroughput             1          100000    ss   25  0.478 ± 0.019   s/op
MessageThroughputTest.measureMessageThroughput             1         1000000    ss   25  4.551 ± 0.029   s/op
MessageThroughputTest.measureMessageThroughput            10            1000    ss   25  0.004 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput            10           10000    ss   25  0.022 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput            10          100000    ss   25  0.210 ± 0.031   s/op
MessageThroughputTest.measureMessageThroughput            10         1000000    ss   25  2.114 ± 0.197   s/op
MessageThroughputTest.measureMessageThroughput           100            1000    ss   25  0.006 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput           100           10000    ss   25  0.022 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput           100          100000    ss   25  0.208 ± 0.012   s/op
MessageThroughputTest.measureMessageThroughput           100         1000000    ss   25  2.527 ± 0.125   s/op
MessageThroughputTest.measureMessageThroughput          1000            1000    ss   25  0.009 ± 0.004   s/op
MessageThroughputTest.measureMessageThroughput          1000           10000    ss   25  0.027 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput          1000          100000    ss   25  0.248 ± 0.009   s/op
MessageThroughputTest.measureMessageThroughput          1000         1000000    ss   25  2.510 ± 0.099   s/op
```

Results with 32 threads:

```
Benchmark                                       (actorCount)  (messageCount)  Mode  Cnt  Score   Error  Units
MessageThroughputTest.measureMessageThroughput             1            1000    ss   25  0.010 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput             1           10000    ss   25  0.070 ± 0.014   s/op
MessageThroughputTest.measureMessageThroughput             1          100000    ss   25  0.538 ± 0.017   s/op
MessageThroughputTest.measureMessageThroughput             1         1000000    ss   25  4.983 ± 0.064   s/op
MessageThroughputTest.measureMessageThroughput            10            1000    ss   25  0.007 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput            10           10000    ss   25  0.030 ± 0.004   s/op
MessageThroughputTest.measureMessageThroughput            10          100000    ss   25  0.313 ± 0.050   s/op
MessageThroughputTest.measureMessageThroughput            10         1000000    ss   25  3.389 ± 0.241   s/op
MessageThroughputTest.measureMessageThroughput           100            1000    ss   25  0.007 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput           100           10000    ss   25  0.039 ± 0.007   s/op
MessageThroughputTest.measureMessageThroughput           100          100000    ss   25  0.367 ± 0.067   s/op
MessageThroughputTest.measureMessageThroughput           100         1000000    ss   25  3.702 ± 0.354   s/op
MessageThroughputTest.measureMessageThroughput          1000            1000    ss   25  0.009 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput          1000           10000    ss   25  0.041 ± 0.005   s/op
MessageThroughputTest.measureMessageThroughput          1000          100000    ss   25  0.340 ± 0.086   s/op
MessageThroughputTest.measureMessageThroughput          1000         1000000    ss   25  3.772 ± 0.369   s/op
```

And finally, back with 16 threads, but capping actors to 16:

```
Benchmark                                       (actorCount)  (messageCount)  Mode  Cnt  Score   Error  Units
MessageThroughputTest.measureMessageThroughput             1            1000    ss   25  0.007 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput             1           10000    ss   25  0.060 ± 0.015   s/op
MessageThroughputTest.measureMessageThroughput             1          100000    ss   25  0.504 ± 0.017   s/op
MessageThroughputTest.measureMessageThroughput             1         1000000    ss   25  5.062 ± 0.064   s/op
MessageThroughputTest.measureMessageThroughput            10            1000    ss   25  0.004 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput            10           10000    ss   25  0.023 ± 0.003   s/op
MessageThroughputTest.measureMessageThroughput            10          100000    ss   25  0.208 ± 0.016   s/op
MessageThroughputTest.measureMessageThroughput            10         1000000    ss   25  2.212 ± 0.126   s/op
MessageThroughputTest.measureMessageThroughput            16            1000    ss   25  0.007 ± 0.001   s/op
MessageThroughputTest.measureMessageThroughput            16           10000    ss   25  0.024 ± 0.002   s/op
MessageThroughputTest.measureMessageThroughput            16          100000    ss   25  0.222 ± 0.019   s/op
MessageThroughputTest.measureMessageThroughput            16         1000000    ss   25  2.242 ± 0.119   s/op
```

Not surprisingly, adding more threads in fact slows things down (context switching). Capping threads to core, with a CPU bound workload, more actors don't help either - in fact, it slows us down a little (but not much, so overhead is minimal). I suspect with I/O bound or heterogeneous workloads this would be different, of course. All that to say, our actors are indeed lightweight.

The biggest issue I saw was that if there's any congestion at any point, you are under very high risk to use up all your memory since the actor inbox/queue is unbounded, and we close over whatever state was in the lambda (e.g. a user request), which could be large.

So say the gateway is sending, idk, 100 commands per second. All good. Suddenly, the brokers are unavailable! Since we have no back pressure on the gateway, it will buffer all requests as actor jobs, and quickly run out of memory.

**Comment by npepinpe** [comment]
Discussed with @deepthidevaki, and I'll try two things:

1. Micro-benchmark to test whether having one actor (BrokerClient) processing lots of messages versus multiple actors processing few is now creating contention for the threads, as opposed to two actors both process lots of messages.
2. Will recreate the benchmarks with large clusters to simulate tons of messages without creating high load - our gut feeling is high load creates a lot of extra issues in Zeebe making results unreliable.

**Comment by npepinpe** [comment]
@deepthidevaki - as I'm going on FTO for the next two weeks, feel free to make any adjustments as you see fit to merge this :+1:

**Review by deepthidevaki — APPROVED** [review]
:rocket: Great job evaluating both cases. 

I have only one small question. So I will merge this PR.

bors merge

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/stream/StreamJobsHandler.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    private ClientStreamId streamId;
+
+    @VisibleForTesting("Allow unit testing behavior")
+    AsyncJobStreamRemover(final ClientStreamer<JobActivationProperties> jobStreamer) {
+      this.jobStreamer = jobStreamer;
+    }
+
+    @Override
+    public void run() {
+      LockUtil.withLock(lock, this::lockedRemove);
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply gateway/src/main/java/io/camunda/zeebe/gateway/impl/stream/StreamJobsHandler.java:173]: Actually, yes :sweat_smile:

**Pull request merged**